sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,635 @@
+  return str;
+}
+
+void ParseHttpVersion(
+    const std::string& httpVersion,
+    uint16_t* majorVersion,
+    uint16_t* minorVersion)
+{
+  auto httpVersionEnd = httpVersion.data() + httpVersion.size();
+
+  // Set response code and HTTP version (i.e. HTTP/1.1)
+  auto majorVersionStart
+      = httpVersion.data() + 5; // HTTP = 4, / = 1, moving to 5th place for version
+  auto majorVersionEnd = std::find(majorVersionStart, httpVersionEnd, '.');
+  auto majorVersionInt = std::stoi(std::string(majorVersionStart, majorVersionEnd));
+
+  auto minorVersionStart = majorVersionEnd + 1; // start of minor version
+  auto minorVersionInt = std::stoi(std::string(minorVersionStart, httpVersionEnd));
+
+  *majorVersion = (uint16_t)majorVersionInt;
+  *minorVersion = (uint16_t)minorVersionInt;
+}
+
+/**
+ * @brief Add a list of HTTP headers to the @RawResponse.
+ *
+ * @remark The \p headers must contain valid header name characters (RFC 7230).
+ * @remark Header name, value and delimiter are expected to be in \p headers.
+ *
+ * @param headers The complete list of headers to be added, in the form "name:value\0",
+ * terminated by "\0".
+ *
+ * @throw if \p headers has an invalid header name or if the delimiter is missing.
+ */
+void AddHeaders(std::string const& headers, std::unique_ptr<RawResponse>& rawResponse)
+{
+  auto begin = headers.data();
+  auto end = begin + headers.size();
+
+  while (*begin != '\0')
+  {
+    auto delimiter = std::find(begin, end, '\0');
+    if (delimiter < end)
+    {
+      rawResponse->AddHeader(
+          reinterpret_cast<uint8_t const*>(begin), reinterpret_cast<uint8_t const*>(delimiter));
+    }
+    else
+    {
+      break;
+    }
+    begin = delimiter + 1;
+  }
+}
+
+} // namespace
+
+void GetErrorAndThrow(const std::string& exceptionMessage)
+{
+  DWORD error = GetLastError();
+  throw Azure::Core::Http::TransportException(
+      exceptionMessage + " Error Code: " + std::to_string(error) + ".");
+}
+
+void WinHttpTransport::GetSessionHandle(std::unique_ptr<Details::HandleManager>& handleManager)
+{
+  // Use WinHttpOpen to obtain a session handle.
+  // The dwFlags is set to 0 - all WinHTTP functions are performed synchronously.
+  // TODO: Use specific user-agent or application name.
+  handleManager->m_sessionHandle = WinHttpOpen(
+      L"WinHTTP Azure SDK",
+      WINHTTP_ACCESS_TYPE_NO_PROXY,
+      WINHTTP_NO_PROXY_NAME,
+      WINHTTP_NO_PROXY_BYPASS,
+      0);
+
+  if (!handleManager->m_sessionHandle)
+  {
+    // Errors include:
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_NOT_ENOUGH_MEMORY
+    GetErrorAndThrow("Error while getting a session handle.");
+  }
+}
+
+void WinHttpTransport::GetConnectionHandle(std::unique_ptr<Details::HandleManager>& handleManager)
+{
+  // TODO: Get port from Url
+  // Specify an HTTP server.
+  // Uses port 80 for HTTP and port 443 for HTTPS.
+  // This function always operates synchronously.
+  handleManager->m_connectionHandle = WinHttpConnect(
```

Comment:
> It looks like the manager is currently only a grouping container but not actually a manager.

That's what I was using this type for, primarily for simpler cleanup of the handles rather than requiring explicit clean up before each throw. Maybe manager isn't the best name for it. That said, it is the one responsible for cleanup. Maybe `HandleLifetimeManager`?